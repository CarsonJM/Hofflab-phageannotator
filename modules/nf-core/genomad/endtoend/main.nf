process GENOMAD_ENDTOEND {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/genomad:1.8.0--pyhdfd78af_0':
        'biocontainers/genomad:1.8.0--pyhdfd78af_0' }"

    input:
    tuple val(meta) , path(fasta)
    path(genomad_db)

    output:
    tuple val(meta), path("${prefix}_virus_summary.tsv")                , emit: virus_summary
    tuple val(meta), path("${prefix}_virus.fna.gz")                     , emit: virus_fasta
    tuple val(meta), path("${prefix}_aggregated_classification.tsv")    , emit: agg_class
    tuple val(meta), path("${prefix}_taxonomy.tsv")                     , emit: taxonomy
    tuple val(meta), path("${prefix}_genes.tsv")                        , emit: genes
    path "versions.yml"                                                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    filename = fasta.toString().tokenize('.')[0]
    """
    genomad \\
        end-to-end \\
        ${fasta} \\
        ./ \\
        ${genomad_db} \\
        --threads ${task.cpus} \\
        ${args}

    mv *_summary/*_virus_summary.tsv ${prefix}_virus_summary.tsv
    mv *_aggregated_classification/*_aggregated_classification.tsv ${prefix}_aggregated_classification.tsv
    mv *_annotate/*_taxonomy.tsv ${prefix}_taxonomy.tsv
    mv *_annotate/*_genes.tsv ${prefix}_genes.tsv
    mv *_summary/*_virus.fna ${prefix}_virus.fna
    gzip ${prefix}_virus.fna

    rm -rf ./${filename}_*/*

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        genomad: \$(echo \$(genomad --version 2>&1) | sed 's/^.*geNomad, version //; s/ .*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    filename = "${fasta}"[0..<"${fasta}".lastIndexOf('.')]
    """
    touch ${prefix}_virus_summary.tsv
    touch ${prefix}_aggregated_classification.tsv
    touch ${prefix}_taxonomy.tsv
    touch ${prefix}_genes.tsv
    echo "" | gzip > ${prefix}_virus.fna.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        genomad: \$(echo \$(genomad --version 2>&1) | sed 's/^.*geNomad, version //; s/ .*\$//')
    END_VERSIONS
    """
}
