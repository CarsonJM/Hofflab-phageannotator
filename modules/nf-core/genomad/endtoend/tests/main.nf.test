nextflow_process {

    name "Test Process GENOMAD_ENDTOEND"
    script "../main.nf"
    process "GENOMAD_ENDTOEND"

    setup {
        run("GENOMAD_DOWNLOAD") {
            script "../../download/main.nf"
        }
    }

    test("fasta.gz") {
        when {
            params {
                genomad_args    = "--splits 5 --sensitivity 0.1 --disable-find-proviruses"
            }
            process {
                """
                input[0] = [
                    [ id:'test', assembler:'megahit_single' ], // meta map
                    file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                input[1] = GENOMAD_DOWNLOAD.out.genomad_db
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() },
            )
        }
    }


    test("fasta.gz - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'test', assembler:'megahit_single' ], // meta map
                    file(params.test_data['sarscov2']['genome']['genome_fasta'], checkIfExists: true)
                    ]
                input[1] = GENOMAD_DOWNLOAD.out.genomad_db
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }
    }
}
