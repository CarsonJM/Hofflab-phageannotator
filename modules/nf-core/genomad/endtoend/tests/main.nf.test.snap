{
    "fasta.gz - stub": {
        "content": [
            [
                "versions.yml:md5,17692db4ae585b447fe2897babeec56b"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T09:43:03.994791332"
    },
    "fasta.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_virus_summary.tsv:md5,923b36db276dfee18ac1f1b57cae617e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_virus.fna.gz:md5,483f4a5dfe60171c86ee9b7e6dff908b"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_aggregated_classification.tsv:md5,702aead49cca96c98e43c5f032f60b26"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_taxonomy.tsv:md5,44c6cf5ff4c836f66648457ae8deb0d5"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_genes.tsv:md5,f84d8a9fce3c92c49e78672bea16561f"
                    ]
                ],
                "5": [
                    "versions.yml:md5,17692db4ae585b447fe2897babeec56b"
                ],
                "agg_class": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_aggregated_classification.tsv:md5,702aead49cca96c98e43c5f032f60b26"
                    ]
                ],
                "genes": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_genes.tsv:md5,f84d8a9fce3c92c49e78672bea16561f"
                    ]
                ],
                "taxonomy": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_taxonomy.tsv:md5,44c6cf5ff4c836f66648457ae8deb0d5"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,17692db4ae585b447fe2897babeec56b"
                ],
                "virus_fasta": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_virus.fna.gz:md5,483f4a5dfe60171c86ee9b7e6dff908b"
                    ]
                ],
                "virus_summary": [
                    [
                        {
                            "id": "test",
                            "assembler": "megahit_single"
                        },
                        "test_megahit_single_virus_summary.tsv:md5,923b36db276dfee18ac1f1b57cae617e"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.1"
        },
        "timestamp": "2024-07-24T09:42:30.581537613"
    }
}